source/s3_meta_request.c

```diff
@@ -596,48 +609,36 @@ static void s_s3_meta_request_prepare_request_task(struct aws_task *task, void *
     /* Client owns this event loop group. A cancel should not be possible. */
     AWS_ASSERT(task_status == AWS_TASK_STATUS_RUN_READY);
 
-    int error_code = AWS_ERROR_SUCCESS;
-
     if (!request->always_send && aws_s3_meta_request_has_finish_result(meta_request)) {
-        aws_raise_error(AWS_ERROR_S3_CANCELED);
-        goto dont_send_clean_up;
-    }
-
-    if (vtable->prepare_request(meta_request, request)) {
-        ++request->num_times_prepared;
-        goto dont_send_clean_up;
+        s_s3_prepare_request_payload_callback_and_destroy(payload, AWS_ERROR_S3_CANCELED);
+        return;
     }
 
-    ++request->num_times_prepared;
-
-    aws_s3_add_user_agent_header(meta_request->allocator, request->send_data.message);
-
-    /* Sign the newly created message. */
-    s_s3_meta_request_sign_request(meta_request, request, s_s3_meta_request_request_on_signed, payload);
-
+    /* Kick off the async vtable->prepare_request()
+     * Each subclass has its own implementation of this. */
+    struct aws_future *preparation_future = vtable->prepare_request_async(request);
+    aws_future_register_callback(preparation_future, s_s3_meta_request_on_request_prepared, payload);
     return;
+}
 
-dont_send_clean_up:
-
-    error_code = aws_last_error_or_unknown();
+/* Called after vtable->prepare_request has succeeded or failed. */
+static void s_s3_meta_request_on_request_prepared(struct aws_future *preparation_future, void *user_data) {
+    struct aws_s3_prepare_request_payload *payload = user_data;
+    struct aws_s3_request *request = payload->request;
+    struct aws_s3_meta_request *meta_request = request->meta_request;
 
-    AWS_LOGF_ERROR(
-        AWS_LS_S3_META_REQUEST,
-        "id=%p Could not prepare request %p due to error %d (%s).",
-        (void *)meta_request,
-        (void *)request,
-        error_code,
-        aws_error_str(error_code));
+    int error_code = aws_future_get_error(preparation_future);
+    preparation_future = aws_future_release(preparation_future);
```

Comment:
this is just the thing where release() functions return NULL
we didn't NEED to null out the pointer, but a lot of the aws-c-s3 code nulls out pointers once they're deleted so I stuck with the style

anyway, I changed this, now the future is released along with anything else stored in the async_ctx, and I don't bother nulling out pointers because it's all going to die 2 lines later